### PR TITLE
Adding fields to template graphs

### PR DESCRIPTION
**Pull request by akalburgi**
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/daml-visual/src/DA/Daml/Visual.hs [hunk 1]
```diff
@@ -151,9 +153,32 @@ addCreateChoice :: TemplateChoices -> Map.Map LF.ChoiceName ChoiceDetails -> Cho
 addCreateChoice TemplateChoices {..} lookupData = nodeIdForChoice lookupData tplNameCreateChoice
     where tplNameCreateChoice = LF.ChoiceName $ T.pack $ DAP.renderPretty (headNote "addCreateChoice" (LF.unTypeConName (LF.tplTypeCon template))) ++ "_Create"
 
-constructSubgraphsWithLables :: Map.Map LF.ChoiceName ChoiceDetails -> TemplateChoices -> SubGraph
-constructSubgraphsWithLables lookupData tpla@TemplateChoices {..} = SubGraph nodesWithCreate template
+-- This is copied from PrettyScenarios but depending on SS for visual seems odd
```
[the remark this answers is not in the document]
Comment (cocreature) [reply compiler/damlc/daml-visual/src/DA/Daml/Visual.hs]: I would prefer to reuse the code, either by just depending on it for now or by moving it to some shared location (not quite sure what would be a good place for it, maybe `DA.Daml.LF.Ast.Util`?

File: compiler/damlc/daml-visual/src/DA/Daml/Visual.hs [hunk 2]
```diff
@@ -151,9 +153,32 @@ addCreateChoice :: TemplateChoices -> Map.Map LF.ChoiceName ChoiceDetails -> Cho
...
+
+typeConFields :: LF.Qualified LF.TypeConName -> LF.World -> [T.Text]
+typeConFields qName world = case LF.lookupDataType qName world of
+  Right dataType -> case LF.dataCons dataType of
+    LF.DataRecord re -> concatMap (typeConFieldsNames world) re
+    LF.DataVariant _ -> [""]
+    LF.DataEnum _ -> [""]
+  Left _ -> error "malformed template constructor"
+
+-- Other way is to go from Expr to Qalified a and then use that qualified
```
Comment (cocreature) [inline compiler/damlc/daml-visual/src/DA/Daml/Visual.hs]: I don’t understand what you mean by “Other way”. Does this code implement the other way? If so, then what’s the first way? Or is this an alternative way of implementing the same thing?

File: compiler/damlc/daml-visual/src/DA/Daml/Visual.hs [hunk 3]
```diff
@@ -151,9 +153,32 @@ addCreateChoice :: TemplateChoices -> Map.Map LF.ChoiceName ChoiceDetails -> Cho
...
+tplQulified :: LF.World -> TemplateChoices -> [T.Text]
```
Comment (cocreature) [inline compiler/damlc/daml-visual/src/DA/Daml/Visual.hs]:
```suggestion
tplQualified :: LF.World -> TemplateChoices -> [T.Text]
```

File: compiler/damlc/daml-visual/src/DA/Daml/Visual.hs [hunk 4]
```diff
@@ -151,9 +153,32 @@ addCreateChoice :: TemplateChoices -> Map.Map LF.ChoiceName ChoiceDetails -> Cho
...
+tplQulified wrld tplc = typeConFields qualTpl wrld
```
Comment (cocreature) [inline compiler/damlc/daml-visual/src/DA/Daml/Visual.hs]:
```suggestion
tplQualified wrld tplc = typeConFields qualTpl wrld
```

File: compiler/damlc/daml-visual/src/DA/Daml/Visual.hs (hunk 3 above)
Comment (cocreature) [inline compiler/damlc/daml-visual/src/DA/Daml/Visual.hs]: This function has a very confusing name. It sounds like it somehow returns the qualified name of a template but that’s not what it does. It returns a list of field names. It might be easier to understand this code if you work with `LF.FieldName` everywhere and only convert to `Text` at the very end.

File: compiler/damlc/daml-visual/src/DA/Daml/Visual.hs [hunk 5]
```diff
@@ -151,9 +153,32 @@ addCreateChoice :: TemplateChoices -> Map.Map LF.ChoiceName ChoiceDetails -> Cho
...
+-- This is copied from PrettyScenarios but depending on SS for visual seems odd
+labledField :: T.Text -> T.Text -> T.Text
+labledField fname "" = fname
+labledField fname label = fname <> "." <> label
+
+typeConFieldsNames :: LF.World -> (LF.FieldName, LF.Type) -> [T.Text]
+typeConFieldsNames world (LF.FieldName fName, LF.TConApp tcn _) = map (labledField fName) (typeConFields tcn world)
+typeConFieldsNames _ (LF.FieldName fName, _) = [fName]
+
+typeConFields :: LF.Qualified LF.TypeConName -> LF.World -> [T.Text]
```
Comment (cocreature) [inline compiler/damlc/daml-visual/src/DA/Daml/Visual.hs]: Can this function run into an infinite loop? What happens if I have something like:

```haskell
data X = X { y : Y }
data Y = Y { x : X }
```
Admittedly that is not a useful type in DAML (since it is strict) but if you change it to something like `data X = X { y : Maybe Y }` it becomes a useful type and afaict the only reason why that won’t cause an infinite loop at the moment is that you don’t recurse into variants at all.

File: compiler/damlc/daml-visual/src/DA/Daml/Visual.hs [hunk 6]
```diff
@@ -151,9 +153,32 @@ addCreateChoice :: TemplateChoices -> Map.Map LF.ChoiceName ChoiceDetails -> Cho
...
+typeConFields qName world = case LF.lookupDataType qName world of
```
Comment (cocreature) [inline compiler/damlc/daml-visual/src/DA/Daml/Visual.hs]: It looks like currently you simply flatten all nested records into a single list of fields. Can we display them in a more hierarchical way? Throwing away the structure that the user created doesn’t seem like a nice way to present this.
I could also see this blowing up on large records if you recurse without a limit. Maybe we should make this configurable, e.g., initially have a single parameter for recursing on fields where `0` does not display any fields in the graph (i.e. the behavior without this PR), 1 goes down a single level, 2 recurses once, …

File: compiler/damlc/daml-visual/src/DA/Daml/Visual.hs [hunk 7]
```diff
@@ -151,9 +153,32 @@ addCreateChoice :: TemplateChoices -> Map.Map LF.ChoiceName ChoiceDetails -> Cho
...
+  Right dataType -> case LF.dataCons dataType of
+    LF.DataRecord re -> concatMap (typeConFieldsNames world) re
+    LF.DataVariant _ -> [""]
+    LF.DataEnum _ -> [""]
+  Left _ -> error "malformed template constructor"
+
+-- Other way is to go from Expr to Qalified a and then use that qualified
+tplQulified :: LF.World -> TemplateChoices -> [T.Text]
+tplQulified wrld tplc = typeConFields qualTpl wrld
+    where qualTpl = LF.Qualified LF.PRSelf (modName tplc) (LF.tplTypeCon $ template tplc)
```
[the remark this answers is not in the document]
Comment (cocreature) [reply compiler/damlc/daml-visual/src/DA/Daml/Visual.hs]: Why do you think `PRSelf` is wrong? Seems to make sense to me.

File: compiler/damlc/daml-visual/src/DA/Daml/Visual.hs [hunk 8]
```diff
@@ -183,11 +208,17 @@ choiceDetailsColorCode False = "green"
...
-subGraphEnd tpl = "label=" ++ DAP.renderPretty (LF.tplTypeCon tpl) ++ ";color=" ++ "blue" ++ "\n}"
+subGraphEnd :: SubGraph -> String
+subGraphEnd sg = "label=<" ++ tHeader ++ tTitle ++ tBody  ++ tclose ++ ">" ++ ";color=" ++ "blue" ++ "\n}"
+    where tHeader = "<table align = \"left\" border=\"0\" cellborder=\"0\" cellspacing=\"1\">\n"
+          tTitle =  "<tr><td align=\"center\"><b>" ++  DAP.renderPretty (LF.tplTypeCon $ clusterTemplate sg) ++ "</b></td></tr>"
+          tBody = concatMap fieldTableLine (templateFileds sg)
+          fieldTableLine field = "<tr><td align=\"left\">" ++ T.unpack field  ++ "</td></tr> \n"
+          tclose = "</table>"
 
+-- DAP.renderPretty (LF.tplTypeCon (clusterTemplate sg))
```
Comment (cocreature) [inline compiler/damlc/daml-visual/src/DA/Daml/Visual.hs]: Can we remove this?

File: compiler/damlc/daml-visual/src/DA/Daml/Visual.hs [hunk 9]
```diff
@@ -183,11 +208,17 @@ choiceDetailsColorCode False = "green"
 subGraphBodyLine :: ChoiceDetails -> String
 subGraphBodyLine chc = "n" ++ show (nodeId chc)++ "[label=" ++ DAP.renderPretty (displayChoiceName chc) ++"][color=" ++ choiceDetailsColorCode (consuming chc) ++"]; "
 
-subGraphEnd :: LF.Template -> String
-subGraphEnd tpl = "label=" ++ DAP.renderPretty (LF.tplTypeCon tpl) ++ ";color=" ++ "blue" ++ "\n}"
+subGraphEnd :: SubGraph -> String
+subGraphEnd sg = "label=<" ++ tHeader ++ tTitle ++ tBody  ++ tclose ++ ">" ++ ";color=" ++ "blue" ++ "\n}"
```
Comment (cocreature) [inline compiler/damlc/daml-visual/src/DA/Daml/Visual.hs]: Why did you change the type here? It looks like you still only use the template. Changing the type seems fine in principle (although I’m not sure it helps with anything here either) but I would suggest that you also change `subGraphHeader` in that case. It seems odd to have different types for `subGraphHeader` and `subGraphEnd`.

File: compiler/lsp-tests/src/Main.hs [hunk 10]
```diff
@@ -435,7 +435,7 @@ executeCommandTests run _ = testGroup "execute command"
         Just escapedFp <- pure $ uriToFilePath (main' ^. uri)
         actualDotString :: ExecuteCommandResponse <- LSP.request WorkspaceExecuteCommand $ ExecuteCommandParams
            "daml/damlVisualize"  (Just (List [Aeson.String $ T.pack escapedFp]))
-        let expectedDotString = "digraph G {\ncompound=true;\nrankdir=LR;\nsubgraph cluster_Coin{\nn0[label=Create][color=green]; \nn1[label=Archive][color=red]; \nn2[label=Delete][color=red]; \nlabel=Coin;color=blue\n}\n}\n"
+        let expectedDotString = "digraph G {\ncompound=true;\nrankdir=LR;\nsubgraph cluster_Coin{\nn0[label=Create][color=green]; \nn1[label=Archive][color=red]; \nn2[label=Delete][color=red]; \nlabel=<<table align = \"left\" border=\"0\" cellborder=\"0\" cellspacing=\"1\">\n<tr><td align=\"center\"><b>Coin</b></td></tr><tr><td align=\"left\">owner</td></tr> \n</table>>;color=blue\n}\n}\n"
```
Comment (cocreature) [inline compiler/lsp-tests/src/Main.hs]: Can we get some proper tests in the shake tests that this works like it is supposed to?

File: compiler/damlc/daml-visual/src/DA/Daml/Visual.hs (hunk 6 above)
Comment (cocreature) [reply compiler/damlc/daml-visual/src/DA/Daml/Visual.hs]: Oh I see, I missed the dots. Thanks for the explanation. I’m still a bit worried that this doesn’t scale to large nested records so I think the limit is probably still a good idea but I’m fine with that being added separately.

File: compiler/damlc/daml-visual/src/DA/Daml/Visual.hs [hunk 11]
```diff
@@ -183,11 +208,17 @@ choiceDetailsColorCode False = "green"
...
+    where tHeader = "<table align = \"left\" border=\"0\" cellborder=\"0\" cellspacing=\"1\">\n"
+          tTitle =  "<tr><td align=\"center\"><b>" ++  DAP.renderPretty (LF.tplTypeCon $ clusterTemplate sg) ++ "</b></td></tr>"
+          tBody = concatMap fieldTableLine (templateFileds sg)
```
Comment (cocreature) [inline compiler/damlc/daml-visual/src/DA/Daml/Visual.hs]:
```suggestion
          tBody = concatMap fieldTableLine (templateFields sg)
```

File: compiler/damlc/daml-visual/src/DA/Daml/Visual.hs (hunk 9 above)
Comment (cocreature) [reply compiler/damlc/daml-visual/src/DA/Daml/Visual.hs]: Oh I see, I missed the call to `templateFields` below. Somehow my browser search failed me :(

File: compiler/lsp-tests/src/Main.hs (hunk 10 above)
Comment (cocreature) [reply compiler/lsp-tests/src/Main.hs]: Just testing a bunch of different case and making sure (without golden tests) that they all behave as you expect, e.g., records, nested records, variants, …

Comment (cocreature) [reply compiler/lsp-tests/src/Main.hs]: What I am trying to say is that we should have tests that check for this in the shake tests. Given that this change did not change any of the Shake tests, they evidently don’t test your change here so we should extend them to test that fields are correctly represented in the graph.

File: compiler/damlc/daml-visual/src/DA/Daml/Visual.hs (hunk 7 above)
[the remark this answers is not in the document]
Comment (cocreature) [reply compiler/damlc/daml-visual/src/DA/Daml/Visual.hs]: As far as I can tell, you only rewrite PRSelf in external packages see https://github.com/digital-asset/daml/blob/master/compiler/damlc/daml-visual/src/DA/Daml/Visual.hs#L127. `pkg` itself can still contain `PRSelf`. As far as I can tell, your only search for templates in the current package not the external package so `PRSelf` should be accurate.

File: compiler/lsp-tests/src/Main.hs (hunk 10 above)
Comment (akalburgi) [reply compiler/lsp-tests/src/Main.hs]: I have updated the tests to assert the fields that were added in this PR. The bit that is not tested in these ShakeIdeTest is the style (the table one which shows template data fields within the DOT file).

File: compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs [hunk 12]
```diff
@@ -538,33 +538,34 @@ timedSection targetDiffTime block = do
...
-graphTest :: LF.World -> LF.Package -> Set.Set TemplateProp -> ShakeTest ()
-graphTest wrld lfPkg expectedProps = do
-    let actual = Set.fromList $ map templateChoicesToProps tplPropsActual
-        tplPropsActual = concatMap (V.moduleAndTemplates wrld) (NM.toList $ LF.packageModules lfPkg)
-    unless (expectedProps == actual) $
-        throwError $ ExpectedTemplateProps expectedProps actual
-
-expectedTemplatePoperties :: D.NormalizedFilePath -> Set.Set TemplateProp -> ShakeTest ()
-expectedTemplatePoperties damlFilePath expectedProps = do
+expectedSubgraph :: V.SubGraph -> ExpectedSubGraph
```
Comment (cocreature) [inline compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs]: Your names are inconsistent: `expectedSubgraph` but `graphToExpectedGraph` instead of `expectedGraph` which also exists but does something different.

File: compiler/damlc/daml-visual/src/DA/Daml/Visual.hs (hunk 5 above)
Comment (cocreature) [reply compiler/damlc/daml-visual/src/DA/Daml/Visual.hs]: That comment still applies afaict. At least it deserves a comment and maybe a TODO that we might also want to recurse on variants at some point which will require changes.

File: compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs (hunk 12 above)
Comment (akalburgi) [reply compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs]: Thanks for point that out. Corrected it

File: compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs [hunk 13]
```diff
@@ -104,7 +103,7 @@ data ShakeTestError
     | ExpectedVirtualResourceNote VirtualResource T.Text (Map VirtualResource T.Text)
     | ExpectedNoVirtualResourceNote VirtualResource (Map VirtualResource T.Text)
     | ExpectedNoErrors [D.FileDiagnostic]
-    | ExpectedTemplateProps (Set.Set TemplateProp) (Set.Set TemplateProp)
+    | ExpectedGraphProps ExpectedGraph ExpectedGraph
```
[the remark this answers is not in the document]
Comment (akalburgi) [reply compiler/damlc/daml-ide-core/test/Development/IDE/State/API/Testing.hs]: That was me not being careful. Changed one to be `Graph`

File: compiler/damlc/daml-visual/src/DA/Daml/Visual.hs (hunk 5 above)
Comment (akalburgi) [reply compiler/damlc/daml-visual/src/DA/Daml/Visual.hs]: Yes, I will add a TODO. Will revisit this again soon
